modules/IsoMux/connection/tls_connection.cpp

```diff
@@ -0,0 +1,413 @@
+        case tls::Connection::result_t::want_write:
+            conn->tls_connection->wait_for(read_res, default_timeout_ms);
+            break;
+        case tls::Connection::result_t::timeout:
+            // the MBedTLS code loops on timeout, is_sequence_timeout() is used instead
+            break;
+        case tls::Connection::result_t::closed:
+        default:
+            result = -1;
+            break;
+        }
+
+        if (conn->ctx->is_connection_terminated) {
+            dlog(DLOG_LEVEL_ERROR, "Reading from tcp-socket aborted");
+            conn->tls_connection->shutdown();
+            result = -2;
+        }
+
+        if (::is_sequence_timeout(ts_start, conn->ctx)) {
+            break;
+        }
+
+        if (not read_complete) {
+            break;
+        }
+    }
+
+    return (result < 0) ? result : static_cast<ssize_t>(bytes_read);
+}
+
+ssize_t connection_write(struct v2g_connection* conn, unsigned char* buf, std::size_t count) {
+    assert(conn != nullptr);
+    assert(conn->tls_connection != nullptr);
+
+    ssize_t result{0};
+    std::size_t bytes_written{0};
+
+    while ((bytes_written < count) && (result >= 0)) {
+        const std::size_t remaining = count - bytes_written;
+        std::size_t bytes_out{0};
+        const auto* ptr = reinterpret_cast<std::byte*>(&buf[bytes_written]);
+
+        const auto write_res = conn->tls_connection->write(ptr, remaining, bytes_out);
+        switch (write_res) {
+        case tls::Connection::result_t::success:
+            bytes_written += bytes_out;
+            break;
+        case tls::Connection::result_t::want_read:
+        case tls::Connection::result_t::want_write:
+            conn->tls_connection->wait_for(write_res, default_timeout_ms);
+            break;
+        case tls::Connection::result_t::timeout:
+            // the MBedTLS code loops on timeout
+            break;
+        case tls::Connection::result_t::closed:
+        default:
+            result = -1;
+            break;
+        }
+    }
+
+    if ((result == -1) && (conn->tls_connection->state() == tls::Connection::state_t::closed)) {
+        // if the connection has closed - return the number of bytes sent
+        result = 0;
+    }
+
+    return (result < 0) ? result : static_cast<ssize_t>(bytes_written);
+}
+
+int connection_proxy(struct v2g_connection* conn, int proxy_fd) {
+
+    dlog(DLOG_LEVEL_INFO, "Multiplexer: Proxy TLS->TCP");
+    int ev_fd = conn->tls_connection->socket(); // underlying socket of TLS connection
+
+    // SupportedAppProtocolReq message is still in buffer, we need to forward it to the external stack
+    write(proxy_fd, conn->buffer, conn->payload_len + 8);
+
+    struct pollfd poll_list[2];
+    poll_list[0].fd = proxy_fd;
+    poll_list[1].fd = ev_fd;
+    poll_list[0].events = POLLIN;
+    poll_list[1].events = POLLIN;
+
+    unsigned char buf[2048];
+
+    // Set reading to (more or less) non-blocking
+    conn->tls_connection->set_read_timeout(10);
```

Comment:
TLS code fully supports non-blocking IO. Used when timeout == 0.
both read() and write() have methods that take the timeout parameter.